timemachine/maps/terminal_bonds.py

```diff
@@ -0,0 +1,236 @@
+"""Summarize HarmonicBonds as Intervals in R+, construct maps between Intervals, & apply these maps to bond lengths"""
+
+from dataclasses import dataclass
+from typing import Tuple
+
+import numpy as np
+from jax import config, jacobian, jit
+from jax import numpy as jnp
+from jax import vmap
+from numpy.typing import NDArray as Array
+
+config.update("jax_enable_x64", True)
+
+from typing import List
+
+import networkx as nx
+
+from timemachine.constants import BOLTZ, DEFAULT_TEMP
+
+
+@dataclass
+class Interval:
+    lower: float
+    upper: float
+
+    @property
+    def width(self) -> float:
+        return self.upper - self.lower
+
+    def __post_init__(self):
+        assert self.width > 0
+        assert self.lower > 0
+
+
+@dataclass
+class Gaussian:
+    mean: float
+    stddev: float
+
+    def to_interval(self, sigma_thresh=20) -> Interval:
+        r = self.stddev * sigma_thresh
+        interval = Interval(self.mean - r, self.mean + r)
+        return interval
+
+    @classmethod
+    def from_harmonic_bond(cls, force_constant, eq_length, temperature=DEFAULT_TEMP):
+        return cls(eq_length, np.sqrt(BOLTZ * temperature / force_constant))
+
+
+@jit
+def interval_map(x, src_lb, src_ub, dst_lb, dst_ub):
+    """map x in Interval[src_lb, src_ub] to y in Interval[dst_lb, dst_ub]"""
+
+    scale_factor = (dst_ub - dst_lb) / (src_ub - src_lb)
+
+    in_support = (x >= src_lb) & (x <= src_ub)
+    mapped = dst_lb + (x - src_lb) * scale_factor
+
+    return jnp.where(in_support, mapped, np.nan)
+
+
+def conf_map(x, bond, param) -> Tuple[Array, Array]:
+    """Apply map to a single bond in a conformer x, return (updated x, logdetjac)"""
+    a, b = bond
+    dim = 3
+
+    def f(xy, param):
+        """R^[2 x dim] -> R^[2 x dim]"""
+        x, y = xy[:dim], xy[dim:]
+        src_lb, src_ub, dst_lb, dst_ub = param
+
+        r = jnp.linalg.norm(x - y)
+        new_r = interval_map(r, src_lb, src_ub, dst_lb, dst_ub)
+
+        vec = (y - x) / jnp.linalg.norm(y - x)
+        y_prime = x + new_r * vec
+
+        xy_prime = jnp.hstack([x, y_prime])
+        return xy_prime
+
+    def map_and_logdetjac(x, y, param):
+        xy = jnp.hstack([x, y])
+        xy_prime = f(xy, param)
+        y_prime = xy_prime[dim:]
+
+        jac = jacobian(f)(xy, param)
+        sign, logdet = jnp.linalg.slogdet(jac)
+
+        # negative det would be unexpected --> nan-poison if detected
+        logdetjac = jnp.where(sign == 1, logdet, jnp.nan)
+
+        return y_prime, logdetjac
+
+    x_b_mapped, logdetjac = map_and_logdetjac(x[a], x[b], param)
+    x_updated = x.at[b].set(x_b_mapped)
+
+    return x_updated, logdetjac
+
+
+apply_conf_map_to_traj = jit(vmap(conf_map, in_axes=(0, None, None)))
+
+
+def apply_conf_maps_to_traj(xs, bond_idxs, params) -> Tuple[Array, Array]:
+    """Apply maps to several bonds in a trajectory of conformers xs, return (updated xs, logdetjacs)"""
+    xs_traj, logdetjac_increments_traj = [jnp.array(xs)], [np.zeros(len(xs))]
```

Comment:
Does xs_traj need to be a list?

Reply:
No -- could switch to a fixed-size array with leading dimension = `len(bond_idxs)` if that would be preferable

Reply:
I meant just replace it with a variable equal to the last value:
```
    xs_updated = jnp.array(xs)
    for (bond, param) in zip(bond_idxs, params):  # TODO: jax.lax for-loop?
        xs_updated, logdetjac_increments = apply_conf_map_to_traj(xs_updated, bond, param)
        logdetjac_increments_traj.append(logdetjac_increments)

    return xs_updated, np.sum(logdetjac_increments_traj, axis=0)
```

Reply:
Ahh, I misunderstood -- applied this change in https://github.com/proteneer/timemachine/pull/801/commits/652211ab83c4f24e6b7900382920e122003e06e2